File: lib/cpp/hawkes/simulation/hawkes_kernels/hawkes_kernel_power_law.cpp
// License: BSD 3 clause


#include "tick/hawkes/simulation/hawkes_kernels/hawkes_kernel_power_law.h"

HawkesKernelPowerLaw::HawkesKernelPowerLaw(double multiplier,
                                           double cutoff,
                                           double exponent,
                                           double support,
                                           double error)
  : HawkesKernel(support),
    multiplier(multiplier), exponent(exponent), cutoff(cutoff) {
  if (support <= 0) {
    if (error <= 0) {
      throw std::invalid_argument("Either support or error must be non negative");
    }
    this->support = pow(error, -1 / exponent) / multiplier - cutoff;
  }
}

HawkesKernelPowerLaw::HawkesKernelPowerLaw()
  : HawkesKernelPowerLaw(0.0, 0.0, 0.0) {
}

double HawkesKernelPowerLaw::get_value_(double x) {
  return multiplier * pow(x + cutoff, -exponent);
}

double HawkesKernelPowerLaw::get_norm(int nsteps) {
  double A = support;
  double norm =
    (pow(A + cutoff, 1 - exponent) - pow(cutoff, 1 - exponent)) * multiplier / (1 - exponent);
  return norm;
}

File: tick/hawkes/simulation/hawkes_kernels/hawkes_kernel_power_law.py
# License: BSD 3 clause

from tick.hawkes.simulation.build.hawkes_simulation import (
    HawkesKernelPowerLaw as _HawkesKernelPowerLaw)
from .hawkes_kernel import HawkesKernel


class HawkesKernelPowerLaw(HawkesKernel):
    """Hawkes kernel for power law

    .. math::
        \\phi(t) = \\phi(t) = \\alpha (\\delta + t)^{- \\beta} 1_{t > 0}

    Where :math:`\\alpha` is called the multiplier, `\\delta` the cut-off and
    :math:`\\beta` the exponent

    Parameters
    ----------
    multiplier : `float`
        Multiplier of the kernel, also noted :math:`\\alpha`

    cutoff : `float`
        Cut-off of the kernel, also noted :math:`\\delta`

    exponent : `float`
        Exponent of the kernel, also noted :math:`\\beta`
    """

    def __init__(self, multiplier, cutoff, exponent, support=-1, error=1e-5):
        HawkesKernel.__init__(self)
        self._kernel = _HawkesKernelPowerLaw(multiplier, cutoff, exponent,
                                             support, error)

    @property
    def multiplier(self):
        return self._kernel.get_multiplier()

    @property
    def cutoff(self):
        return self._kernel.get_cutoff()

    @property
    def exponent(self):
        return self._kernel.get_exponent()

    def __str__(self):
        if self.multiplier == 0:
            return '0'
        elif self.exponent == 0:
            return '{:g}'.format(self.multiplier)
        else:
            return '{:g} * ({:g} + t)^(-{:g})'.format(
                self.multiplier, self.cutoff, self.exponent)

    def __repr__(self):
        return self.__str__().replace(' ', '')

    def __strtex__(self):
        if self.multiplier == 0:
            return r'$0$'
        elif self.exponent == 0:
            return r'${:g}$'.format(self.multiplier)
        else:
            if self.multiplier == 1:
                return r'$(%g+t)^{-%g}$' % (self.cutoff, self.exponent)
            else:
                return r'$%g (%g+t)^{-%g}$' % (self.multiplier, self.cutoff,
                                               self.exponent)
